File: day2/day2.rb
lines = open('day2.txt').each_line
  .map { _1 =~ /^(\d+)-(\d+) (.): / and [$', $3, $1.to_i, $2.to_i] }
  .compact

part1 = lines.count { (_3.._4).include? _1.count(_2) }
part2 = lines.count { (_1[_3 - 1] == _2) ^ (_1[_4 - 1] == _2) }

puts "Part1: #{part1}"
puts "Part1: #{part2}"

File: 2021/day02/day2.hs
module Main where
import System.IO  
import Control.Monad
import Data.Foldable

readLines :: String -> IO [String]
readLines file = lines <$> readFile file

classify :: String -> (Char, Int)
classify line = (head wl, read wr)
   where (wl:wr:_) = words line


part1 :: [String] -> Int
part1 ls = h*d
   where
      (h,d) = foldl' go (0,0) (classify <$> ls)
      go (h,d) ('f', x) = (h+x, d)
      go (h,d) ('u', x) = (h, d-x)
      go (h,d) ('d', x) = (h, d+x)

part2 :: [String] -> Int
part2 ls = h*d
   where
      (h,d,_) = foldl' go (0,0,0) (classify <$> ls)
      go (h, d, a) ('f', x) = (h + x, d + a*x, a)
      go (h, d, a) ('d', x) = (h, d, a + x)
      go (h, d, a) ('u', x) = (h, d, a - x)

main :: IO ()
main = do
   contents <- readLines "day2.txt"
   putStrLn $ "part1: " ++ (show $ part1 contents)
   putStrLn $ "part2: " ++ (show $ part2 contents)
